airflow/migrations/versions/83f031fd9f1c_change_ts_columns_to_datetime_on_mssql.py

```diff
@@ -0,0 +1,181 @@
+    """
+    This function return primary and unique constraint
+    along with column name. some tables like task_instance
+    is missing primary key constraint name and the name is
+    auto-generated by sql server. so this function helps to
+    retrieve any primary or unique constraint name.
+
+    :param conn: sql connection object
+    :param table_name: table name
+    :return: a dictionary of ((constraint name, constraint type), column name) of table
+    :rtype: defaultdict(list)
+    """
+    query = """SELECT tc.CONSTRAINT_NAME , tc.CONSTRAINT_TYPE, ccu.COLUMN_NAME
+     FROM INFORMATION_SCHEMA.TABLE_CONSTRAINTS AS tc
+     JOIN INFORMATION_SCHEMA.CONSTRAINT_COLUMN_USAGE AS ccu ON ccu.CONSTRAINT_NAME = tc.CONSTRAINT_NAME
+     WHERE tc.TABLE_NAME = '{table_name}' AND
+     (tc.CONSTRAINT_TYPE = 'PRIMARY KEY' or UPPER(tc.CONSTRAINT_TYPE) = 'UNIQUE')
+    """.format(
+        table_name=table_name
+    )
+    result = conn.execute(query).fetchall()
+    constraint_dict = defaultdict(list)
+    for constraint, constraint_type, column in result:
+        constraint_dict[(constraint, constraint_type)].append(column)
+    return constraint_dict
+
+
+def drop_column_constraints(operator, column_name, constraint_dict):
+    """
+    Drop a primary key or unique constraint
+
+    :param operator: batch_alter_table for the table
+    :param constraint_dict: a dictionary of ((constraint name, constraint type), column name) of table
+    """
+    for constraint, columns in constraint_dict.items():
+        if column_name in columns:
+            if constraint[1].lower().startswith("primary"):
+                operator.drop_constraint(constraint[0], type_='primary')
+            elif constraint[1].lower().startswith("unique"):
+                operator.drop_constraint(constraint[0], type_='unique')
+
+
+def create_constraints(operator, column_name, constraint_dict):
+    """
+    Create a primary key or unique constraint
+
+    :param operator: batch_alter_table for the table
+    :param constraint_dict: a dictionary of ((constraint name, constraint type), column name) of table
+    """
+    for constraint, columns in constraint_dict.items():
+        if column_name in columns:
+            if constraint[1].lower().startswith("primary"):
+                operator.create_primary_key(constraint_name=constraint[0], columns=columns)
+            elif constraint[1].lower().startswith("unique"):
+                operator.create_unique_constraint(constraint_name=constraint[0], columns=columns)
+
+
+def __use_date_time2(conn):
+    result = conn.execute(
+        """SELECT CASE WHEN CONVERT(VARCHAR(128), SERVERPROPERTY ('productversion'))
+        like '8%' THEN '2000' WHEN CONVERT(VARCHAR(128), SERVERPROPERTY ('productversion'))
+        like '9%' THEN '2005' ELSE '2005Plus' END AS MajorVersion"""
+    ).fetchone()
+    mssql_version = result[0]
+    return mssql_version not in ("2000", "2005")
+
+
+def __is_timestamp(conn, table_name, column_name):
```

Comment:
Any reason for two underscore -- in this migration and the last?

```suggestion
def _is_timestamp(conn, table_name, column_name):
```

Reply:
have fixed it